src/core/lib/security/transport/client_auth_filter.cc

```diff
@@ -119,34 +120,48 @@ static void on_credentials_metadata(grpc_exec_ctx *exec_ctx, void *arg,
   }
 }
 
-void build_auth_metadata_context(grpc_security_connector *sc,
-                                 grpc_auth_context *auth_context,
-                                 call_data *calld) {
-  char *service = grpc_slice_to_c_string(calld->method);
+void grpc_auth_metadata_context_build(
+    const char *url_scheme, grpc_slice call_host, grpc_slice call_method,
+    grpc_auth_context *auth_context,
+    grpc_auth_metadata_context *auth_md_context) {
+  char *service = grpc_slice_to_c_string(call_method);
   char *last_slash = strrchr(service, '/');
   char *method_name = NULL;
   char *service_url = NULL;
-  reset_auth_metadata_context(&calld->auth_md_context);
+  grpc_auth_metadata_context_reset(auth_md_context);
   if (last_slash == NULL) {
     gpr_log(GPR_ERROR, "No '/' found in fully qualified method name");
     service[0] = '\0';
+    method_name = gpr_strdup("");
   } else if (last_slash == service) {
-    /* No service part in fully qualified method name: will just be "/". */
-    service[1] = '\0';
+    method_name = gpr_strdup("");
   } else {
     *last_slash = '\0';
     method_name = gpr_strdup(last_slash + 1);
   }
-  if (method_name == NULL) method_name = gpr_strdup("");
-  char *host = grpc_slice_to_c_string(calld->host);
-  gpr_asprintf(&service_url, "%s://%s%s",
-               sc->url_scheme == NULL ? "" : sc->url_scheme, host, service);
-  calld->auth_md_context.service_url = service_url;
-  calld->auth_md_context.method_name = method_name;
-  calld->auth_md_context.channel_auth_context =
+  char *host_and_port = grpc_slice_to_c_string(call_host);
+  if (strcmp(url_scheme, GRPC_SSL_URL_SCHEME) == 0) {
+    /* Remove the port if it is 443. */
+    char *host;
+    char *port;
+    if (gpr_split_host_port(host_and_port, &host, &port)) {
```

Comment:
It seems unnecessary to use `gpr_split_host_port()` here, since we're just going to wind up using `strrchr()` anyway.  This results in doing the same split twice and allocating memory without any real need.

Instead, I suggest something like this:

```
char *colon = strrchr(host_and_port, ':');
if (colon != NULL && strcmp(colon + 1, "443") == 0) {
  *colon = '\0';
}
```

Reply:
Thanks,

I used the split_host_port to take care of IPV6 literals like these:
`[1080:0:0:0:8:800:200C:443]`
On the other hand, you're right that the strcmp would not work here ("443]" vs "443"). I'll probably fix it then since calling `gpr_split_host_port` results in 2 new mallocs...

Reply:
If you did want this check to work for the IPv6 literal case, perhaps you could use something like this:

```
strncmp(colon + 1, "443", 3) && (colon[4] == '\0' || colon[4] == ']')
```

In any case, this change LGTM.